### PR TITLE
[AIRFLOW-850] Add a PythonSensor

### PR DESCRIPTION
**Pull request by feng-tao**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-850
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
A general purpose PythonSensor which allows an arbitrary Python callable to delay Task execution until the callable returns True. This is based on a stale pr(https://github.com/apache/incubator-airflow/pull/2058)

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
yes

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by feng-tao** [comment]
PTAL @ashb  @Fokko , this is based on the stale pr (https://github.com/apache/incubator-airflow/pull/2058) which I think is a good addition to Airflow sensor.

File: airflow/contrib/sensors/python_sensor.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+# specific language governing permissions and limitations
+# under the License.
+
+from airflow.sensors.base_sensor_operator import BaseSensorOperator
+from airflow.utils.decorators import apply_defaults
+
+
+class PythonSensor(BaseSensorOperator):
+    """
+    Waits for a Python callable to return True
```
Comment (ashb) [inline airflow/contrib/sensors/python_sensor.py]: Adding a short example here in the docs specifying what arguments are passed to the callable, how to access templates_ditc etc would be good I think

File: airflow/contrib/sensors/python_sensor.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        self.op_args = op_args or []
+        self.op_kwargs = op_kwargs or {}
+        self.provide_context = provide_context
+        self.templates_dict = templates_dict
+
+    def poke(self, context):
+        if self.provide_context:
+            context.update(self.op_kwargs)
+            context['templates_dict'] = self.templates_dict
+            self.op_kwargs = context
```
Comment (ashb) [inline airflow/contrib/sensors/python_sensor.py:77]: Does it matter that we end up doing this each time we poke?

Comment (feng-tao) [reply airflow/contrib/sensors/python_sensor.py:77]: I think we need to have a way to access the context for the sensor. Not sure if there is a better way to do so. Besides, I think it is ok to do that as it is not an expensive operation.

File: airflow/contrib/sensors/python_sensor.py (hunk 1 above)
Comment (feng-tao) [reply airflow/contrib/sensors/python_sensor.py]: updated

**Comment by feng-tao** [comment]
hey @ashb , update the pr with an example. PTAL.

**Review by kaxil — APPROVED** [review]

**Pull request merged**